text/0020-block-proposer-improvements.md

```diff
@@ -0,0 +1,108 @@
+
+The solution for this issue could be to include the hash of the last finalized block in the request for deploys. The Block Proposer could then check if it already has that block as finalized, and if not, it would buffer the request and complete it when it receives the notification of the relevant block's finalization.
+
+Even with this solution in place, we could actually still experience problems. Consensus could finalize multiple blocks at once, and if the events aren't guaranteed to arrive in order, the Block Proposer could be notified of the finalization of the last block before some others. In such a case, it would still have an incomplete picture of the new block's past. To prevent this, we would have to only handle notifications about finalized blocks if the block's parent has already been finalized (and buffer finalization notifications that don't satisfy this condition, until it is satisfied).
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+### Seeing finalization before proposal
+
+When receiving `FinalizedProtoBlock(block)`:
+
+- check that `block.hash()` is in `self.state.proposed`
+- if not, add `block.hash()` to a new collection `self.state.finalization_queue`
+- if yes, proceed in the current manner
+
+When receiving `ProposedProtoBlock(block)`:
+
+- check if `block.hash()` is in `self.state.finalization_queue`
+- if yes, remove it from `finalization_queue` and proceed as if we received both `ProposedProtoBlock` and `FinalizedProtoBlock`
+- if no, proceed in the current manner
+
+### Request for a block before seeing another validator's proposal
+
+This would require adding a `last_finalized_block: BlockHash` field to `BlockProposerRequest::ListForInclusion`.
+
+When receiving a `ListForInclusion` request:
+
+- check if `last_finalized_block` is in `self.state.finalized`
+- if yes, proceed in the current manner
+- if not, buffer the request in a new field `self.state.request_queue`
+
+When receiving a `FinalizedProtoBlock` event:
+
+- check if the block's parent is in `self.state.finalized` (in addition to other checks that are in place)
+- if yes, proceed; re-check whether blocks in `self.state.finalization_queue` can be finalized and finalize those that can
+- if no, buffer the block in `self.state.finalization_queue`
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+This adds more state to the `BlockProposer`, which might be undesirable. However, the issues caused by out-of-order events are pressing and the benefits likely outweigh the cost.
+
+## Rationale and alternatives
```

Comment:
To my mind, the main alternative is to get rid of the `proposed` collection. I don't understand why we need it anymore.
Because whenever we propose deploys, we now know all ancestors, so we can just exclude those. There's no harm if they are still in the `pending` collection. That would also resolve the issue with finalizing a block that wasn't proposed: The block proposer wouldn't have to care at all about the event that a new block has been proposed. And we wouldn't have to detect orphaned blocks.

So basically we'd just have the `pending` and `finalized` collections (even the latter could just contain deploys, not blocks).
* Whenever we're asked to propose deploys, we take stuff from `pending`, and exclude all ancestors.
* If we receive a new incoming deploy, we add it to `pending` except if it's already in `finalized`.
* If we're told about a newly finalized block we add its deploys to `finalized` and remove them from `pending`. (No error if they aren't there!)
* We also track the last finalized block, as you propose above (and check for its parent etc.).

Reply:
I think one of the main reasons to have `proposed` collection was so that we don't send all deploy hashes from the consensus to the block executor, but refer to them by the proto hash which is shorter and should uniquely identify the deploys.

Reply:
But then we _are_ sending it all deploy hashes, aren't we?
The difference is that currently we send them when we learn about a _proposed_ block, and with my proposal, we would only send them when _finalizing_ a block.

Reply:
The idea here was to compress the data slightly, I think. But the idea was born when we were still considering sending the _complete_ past with every proposal request, and in such a case the list of block hashes should be much more efficient than the list of deploy hashes. However, if we are sending just the not-yet-finalized part of the past, we can just as well make it a list of the deploy hashes, I think. So then we will know that we should include all the deploys that aren't in finalized blocks or in this list - and we don't need to know about which blocks were proposed :thinking:

Reply:
I think the idea has always been to send only the _not-yet-finalized_ part and rely on the fact that block proposer knows the finalized part.

Reply:
Ah, right, good point! A list of block hashes would be smaller, of course.
But it would also require unique identifiers for deploy lists in different blocks, since we don't have block hashes yet at this point.

Reply:
I think with what I'm suggesting above we'd also reduce the surface between the consensus component and the rest: We could get rid of [`announce_proposed_proto_block`](https://github.com/CasperLabs/casper-node/blob/3516f68a5f86db080fb4495a39f58929b5d81f25/node/src/effect.rs#L856). (Which is currently [called too early anyway](https://github.com/CasperLabs/casper-node/blob/3516f68a5f86db080fb4495a39f58929b5d81f25/node/src/components/consensus/era_supervisor.rs#L546), I think: It needs to wait for the evidence for accusations… I made a note of it.)

Reply:
(I created https://github.com/CasperLabs/casper-node/pull/525 to fix that. :point_up_2:)